File: src/sokoban/env.py
"""
Sokoban environment with movement, successor generation and conservative deadlocks.
"""

from collections import deque

from sokoban.state import Level, Position, Push, State


class SokobanEnv:
    def __init__(self, level: Level, debug_deadlocks: bool = False):
        self.level = level
        self.moves: dict[str, tuple[int, int]] = {
            "UP": (-1, 0),
            "DOWN": (1, 0),
            "LEFT": (0, -1),
            "RIGHT": (0, 1),
        }
        self.debug_deadlocks = debug_deadlocks
        self.deadlock_stats: dict[str, int] = {
            "corner": 0,
            "dead_square": 0,
            "2x2": 0,
        }

        self.point_to_goal = self.precompute_point_to_goal()

    # ------------------------------------------------------------------
    # Basic helpers
    # ------------------------------------------------------------------

    def _record_deadlock(self, name: str) -> None:
        if self.debug_deadlocks:
            self.deadlock_stats[name] += 1

    def _is_floor(self, pos: Position) -> bool:
        return (
            0 <= pos.row < self.level.height
            and 0 <= pos.col < self.level.width
            and pos not in self.level.walls
        )

    def _neighbors(self, pos: Position) -> tuple[Position, Position, Position, Position]:
        return (
            Position(pos.row - 1, pos.col),
            Position(pos.row + 1, pos.col),
            Position(pos.row, pos.col - 1),
            Position(pos.row, pos.col + 1),
        )

    def is_goal_state(self, state: State) -> bool:
        return all(box in self.level.goals for box in state.boxes)

    # ------------------------------------------------------------------
    # Reachability / movement
    # ------------------------------------------------------------------

    def _reachable_floor(self, state: State) -> set[Position]:
        reachable = {state.player}
        queue = deque([state.player])
        while queue:
            pos = queue.popleft()
            for nxt in self._neighbors(pos):
                if nxt in reachable or nxt in self.level.walls or nxt in state.boxes:
                    continue
                reachable.add(nxt)
                queue.append(nxt)
        return reachable

    def normalize_player(self, state: State) -> State:
        reachable = self._reachable_floor(state)
        return State(player=min(reachable), boxes=state.boxes)

    def step(self, state: State, push: Push) -> State:
        box_pos, direction = push
        dy, dx = self.moves[direction]
        new_box_pos = Position(box_pos.row + dy, box_pos.col + dx)
        boxes = set(state.boxes)
        boxes.remove(box_pos)
        boxes.add(new_box_pos)
        # after a push, player stands where the pushed box used to be
        return State(player=box_pos, boxes=frozenset(boxes))

    def _valid_pushes_from_reachable(self, reachable: set[Position], boxes: frozenset[Position]) -> list[Push]:
        pushes: list[Push] = []
        for p in reachable:
            for direction, (dy, dx) in self.moves.items():
                box_pos = Position(p.row + dy, p.col + dx)
                if box_pos not in boxes:
                    continue
                dst = Position(box_pos.row + dy, box_pos.col + dx)
                if self._is_floor(dst) and dst not in boxes:
                    pushes.append(Push(box_pos, direction))
        return pushes

    def get_valid_pushes(self, state: State) -> list[Push]:
        return self._valid_pushes_from_reachable(self._reachable_floor(state), state.boxes)

    def get_successors(self, state: State, use_macro: bool = False) -> list[tuple[State, tuple[Push, ...]]]:
        successors: list[tuple[State, tuple[Push, ...]]] = []
        for push in self.get_valid_pushes(state):
            nxt = self.normalize_player(self.step(state, push))
            successors.append((nxt, (push,)))
        return successors

    # ------------------------------------------------------------------
    # Heuristic support precompute
    # ------------------------------------------------------------------

    def precompute_point_to_goal(self) -> dict[Position, dict[Position, int]]:
        point_to_goal: dict[Position, dict[Position, int]] = {
            Position(r, c): {}
            for r in range(self.level.height)
            for c in range(self.level.width)
            if Position(r, c) not in self.level.walls
        }

        for goal in self.level.goals:
            queue = deque([(goal, 0)])
            seen = {goal}
            while queue:
                pos, dist = queue.popleft()
                if pos in point_to_goal:
                    point_to_goal[pos][goal] = dist

                # reverse-push graph
                for dy, dx in self.moves.values():
                    prev_box = Position(pos.row - dy, pos.col - dx)
                    prev_player = Position(pos.row - 2 * dy, pos.col - 2 * dx)
                    if (
                        prev_box not in seen
                        and prev_box not in self.level.walls
                        and prev_player not in self.level.walls
                    ):
                        seen.add(prev_box)
                        queue.append((prev_box, dist + 1))
        return point_to_goal

    # ------------------------------------------------------------------
    # Deadlocks
    # ------------------------------------------------------------------

    def is_deadlock(self, state: State) -> bool:
        checks = (
            ("corner", self.has_corner_deadlock),
            ("dead_square", self.has_dead_square_deadlock),
            ("2x2", self.has_2x2_deadlock),
        )
        for name, fn in checks:
            if fn(state):
                self._record_deadlock(name)
                return True
        return False

    def has_corner_deadlock(self, state: State) -> bool:
        walls = self.level.walls
        goals = self.level.goals
        for box in state.boxes:
            if box in goals:
                continue
            up = Position(box.row - 1, box.col)
            down = Position(box.row + 1, box.col)
            left = Position(box.row, box.col - 1)
            right = Position(box.row, box.col + 1)
            if (up in walls and left in walls) or (up in walls and right in walls):
                return True
            if (down in walls and left in walls) or (down in walls and right in walls):
                return True
        return False

    def has_dead_square_deadlock(self, state: State) -> bool:
        for box in state.boxes:
            if box not in self.level.goals and not self.point_to_goal.get(box):
                return True
        return False

    def has_2x2_deadlock(self, state: State) -> bool:
        blocked = set(state.boxes)
        blocked.update(self.level.walls)
        boxes = state.boxes
        goals = self.level.goals
        for r in range(self.level.height - 1):
            for c in range(self.level.width - 1):
                p00 = Position(r, c)
                p10 = Position(r + 1, c)
                p01 = Position(r, c + 1)
                p11 = Position(r + 1, c + 1)
                if p00 in blocked and p10 in blocked and p01 in blocked and p11 in blocked:
                    block_boxes = [p for p in (p00, p10, p01, p11) if p in boxes]
                    if not block_boxes:
                        continue
                    if all(p not in goals for p in block_boxes):
                        return True
        return False

File: src/sokoban/env.c
#include "env.h"
#include <stdlib.h>
#include <string.h>

static bool is_wall(const Level* level, int r, int c) {
    if (r < 0 || r >= level->height || c < 0 || c >= level->width) return true;
    return level->walls[r * level->width + c];
}

static bool is_goal(const Level* level, int r, int c) {
    if (r < 0 || r >= level->height || c < 0 || c >= level->width) return false;
    return level->goals[r * level->width + c];
}

static bool is_box_at(int r, int c, const State* state) {
    for (int i = 0; i < state->num_boxes; i++) {
        if (state->boxes[i].row == r && state->boxes[i].col == c) return true;
    }
    return false;
}

State reset(const Level* level) {
    State init_copy;
    init_copy.player = level->init_state.player;
    init_copy.num_boxes = level->init_state.num_boxes;
    
    init_copy.boxes = malloc(init_copy.num_boxes * sizeof(Position));
    memcpy(init_copy.boxes, level->init_state.boxes, init_copy.num_boxes * sizeof(Position));
    
    return init_copy;
}

State step(const State* state, Push push) {
    State next_state;
    next_state.player = push.position;
    next_state.num_boxes = state->num_boxes;
    
    next_state.boxes = malloc(state->num_boxes * sizeof(Position));
    memcpy(next_state.boxes, state->boxes, state->num_boxes * sizeof(Position));

    int dr = 0, dc = 0;
    if (strcmp(push.direction, "UP") == 0) dr = -1;
    else if (strcmp(push.direction, "DOWN") == 0) dr = 1;
    else if (strcmp(push.direction, "LEFT") == 0) dc = -1;
    else if (strcmp(push.direction, "RIGHT") == 0) dc = 1;

    for (int i = 0; i < next_state.num_boxes; i++) {
        if (next_state.boxes[i].row == push.position.row && next_state.boxes[i].col == push.position.col) {
            next_state.boxes[i].row += dr;
            next_state.boxes[i].col += dc;
            break;
        }
    }
    return next_state;
}

bool is_goal_state(const State* state, const Level* level) {
    for (int i = 0; i < state->num_boxes; i++) {
        if (!is_goal(level, state->boxes[i].row, state->boxes[i].col)) {
            return false;
        }
    }
    return true;
}

bool is_deadlock(const State* state, const Level* level) {
    int dr[] = {-1, 1, 0, 0};
    int dc[] = {0, 0, -1, 1};

    for (int b = 0; b < state->num_boxes; b++) {
        int r = state->boxes[b].row;
        int c = state->boxes[b].col;
        
        if (is_goal(level, r, c)) continue;

        bool vert_blocked = false, horz_blocked = false;
        for (int i = 0; i < 4; i++) {
            if (is_wall(level, r + dr[i], c + dc[i]) || is_box_at(r + dr[i], c + dc[i], state)) {
                if (i < 2) vert_blocked = true;
                else horz_blocked = true;
            }
        }
        if (horz_blocked && vert_blocked) return true;

        int offsets[4][4][2] = {
            {{0,0}, {0,1}, {1,0}, {1,1}},   {{0,0}, {0,-1}, {1,0}, {1,-1}},
            {{0,0}, {0,1}, {-1,0}, {-1,1}}, {{0,0}, {0,-1}, {-1,0}, {-1,-1}}
        };
        for (int i = 0; i < 4; i++) {
            bool quad_blocked = true, goal_in_quad = false;
            for (int j = 0; j < 4; j++) {
                int qr = r + offsets[i][j][0];
                int qc = c + offsets[i][j][1];
                
                bool is_w = is_wall(level, qr, qc);
                bool is_b = is_box_at(qr, qc, state);
                
                if (!is_w && !is_b) { quad_blocked = false; break; }
                if (is_b && is_goal(level, qr, qc)) goal_in_quad = true;
            }
            if (quad_blocked && !goal_in_quad) return true;
        }

        int wall_checks[4][6] = {
            {-1, 0,   0, -1,   0, 1},
            { 1, 0,   0, -1,   0, 1},
            { 0,-1,  -1,  0,   1, 0},
            { 0, 1,  -1,  0,   1, 0}
        };

        for (int i = 0; i < 4; i++) {
            int w_dr = wall_checks[i][0];
            int w_dc = wall_checks[i][1];

            if (is_wall(level, r + w_dr, c + w_dc)) {
                bool is_trapped = true;

                for (int scan = 0; scan < 2; scan++) {
                    int s_dr = wall_checks[i][2 + (scan * 2)];
                    int s_dc = wall_checks[i][3 + (scan * 2)];
                    
                    int curr_r = r;
                    int curr_c = c;
                    bool safe_in_this_dir = false;

                    while (true) {
                        curr_r += s_dr;
                        curr_c += s_dc;

                        if (is_wall(level, curr_r, curr_c)) {
                            break; 
                        }

                        if (is_goal(level, curr_r, curr_c)) {
                            safe_in_this_dir = true;
                            break;
                        }

                        if (!is_wall(level, curr_r + w_dr, curr_c + w_dc)) {
                            safe_in_this_dir = true;
                            break;
                        }
                    }

                    if (safe_in_this_dir) {
                        is_trapped = false;
                        break; 
                    }
                }

                if (is_trapped) {
                    return true;
                }
            }
        }
    }
    
    return false;
}

Push* get_valid_pushes(const State* state, const Level* level, int* out_count) {
    int grid_size = level->width * level->height;
    
    bool* visited = calloc(grid_size, sizeof(bool));
    Position* queue = malloc(grid_size * sizeof(Position));
    Push* valid_pushes = malloc(state->num_boxes * 4 * sizeof(Push));
    
    *out_count = 0;
    int head = 0, tail = 0;
    
    queue[tail++] = state->player;
    visited[state->player.row * level->width + state->player.col] = true;

    int dr[] = {-1, 1, 0, 0};
    int dc[] = {0, 0, -1, 1};
    const char* dirs[] = {"UP", "DOWN", "LEFT", "RIGHT"};

    while (head < tail) {
        Position curr = queue[head++];
        
        for (int i = 0; i < 4; i++) {
            int nr = curr.row + dr[i];
            int nc = curr.col + dc[i];
            
            if (nr < 0 || nr >= level->height || nc < 0 || nc >= level->width) continue;
            if (is_wall(level, nr, nc) || visited[nr * level->width + nc]) continue;

            if (is_box_at(nr, nc, state)) {
                int b_nr = nr + dr[i];
                int b_nc = nc + dc[i];
                
                if (!is_wall(level, b_nr, b_nc) && !is_box_at(b_nr, b_nc, state)) {
                    valid_pushes[*out_count].position.row = nr;
                    valid_pushes[*out_count].position.col = nc;
                    valid_pushes[*out_count].direction = dirs[i];
                    (*out_count)++;
                }
            } else {
                visited[nr * level->width + nc] = true;
                queue[tail++] = (Position){nr, nc};
            }
        }
    }
    
    free(visited);
    free(queue);
    
    return valid_pushes;
}